<!DOCTYPE html>
<html>
<head>
  <title>My Portfolio</title>
   <style>
    body {
      background-color: black;
      color: white;
      font-family: Arial, sans-serif;
      padding: 20px;
    }
    a {
      color: skyblue;
      text-decoration: none;
    }
    a:hover {
      text-decoration: underline;
    }
    .project {
      margin-bottom: 20px;
    }
  </style>
</head>
<body>
  <h1>My Most Recent Projects</h1>
  <div class="AI - Clinical Informatics">
    <h2><a href="https://github.com/cartiktq/PMS-ClinicalInformatics">AI in Clinical Informatics</a></h2>
    <img align="center" src="PMS-ClinicalInformatics/PMS-ClinicalInformaticsWorkflow.jpg" alt="Workflow for processing and analyzing information from synthezied clinical notes"/>    
    <div>
      <p>
      This is a personal project that is an extension of some of the research work I did at Harvard Medical School. Phelan-McDermid Syndrome (PMS) is a 
      rare disorder that afflicts children. It is associated with the deletion of the terminal end of the long-arm of the 23rd chromosome and presents
      a variety of clinical and behavioral symptoms. 
      </p>
      <p>
      Since clinical data is extremely difficult to obtain, requiring extensive paperwork and vetting by Institutional Review Boards and other 
      organizations, and since misuse or leakage of clinical data carries heavy penalties such as stipulated in regulations such as HIPAA and 
      GDPR, generative AI can offer a solution to test data pre-processing including anonymization on synthetically generated clinical notes. 
      Given the scarcity of structured patient data in regular clinical data warehouses, clinical notes can be a significant source of information 
      for patients diagnosed with PMS. 
      </p>  
      <p> In this project, I demonstrate the utility of agentic AI based workflows which:
        <ul>
          <li>generate synthetic clinical notes containing fake PHI elements along with all mentions of diagnoses, lab test results, therapeutic 
          procedures, prescriptions, clinical and behavioral symptoms and comborbidities</li>
          <li>use an anonymization tool to scrub PHI elements from the clinical notes</li>
          <li>evaluate the effectiveness of the anonymization pipeline</li>
          <li>use classical Named Entity Recognition (NER) on the scrubbed clinical notes to identify clinically relevant concepts</li>
          <li>map the identified concepts to defined concepts with Concept Unique Identifiers (CUIs) from the Unified Medical Language System (UMLS; 
          this requires getting licensed accss to UMLS from the National Library of Medicine)</li>
          <li>find mappings from the CUIs to concepts defined in specialized terminologies such as SNOMED CT, ICD-9/10, CPT, RxNORM, LOINC, and OMIM</li>
          <li>generate lists of RDF like triples in CSV format as a structured representation of the important clinical information associated with each
          patient</li>
          <li>create a patient centric knowledge graph representation of each patient in the cohort</li>
          <li>(possibly) create hyperdimensional embeddings using transformers to store and cluster patients information in a vector database</li>
          <li>perform some simple clustering to stratify the patient cohort on the basis of commonly reported symptoms. These could be useful as
          phenotypical biomarkers to aid the potential identification of undiagnosed patients (PMS can be diagnosed only be genetic tests)</li>
        </ul>
      </p> 
      <p>
        While the generated embedding vectors for each patient can be presented in user interfaces as knowledge graphs, the clusters (patient strata)
        could be, in future work (in progress) be correlated with genetic variants reported by PMS patients and stored in open access databases such
        dbGaP and licensed databases such as SFARI.
      </p>
      <p>
        From a deployment perspective, this is a modular architecture that can be easily containerized (as each module is a self-contained Python class 
        with Gen AI, ML and other Python packages) and invoked as an orchestrated sequence with Kubernetes in a cloud environment such as Amazon EKS. 
      </p>
    </div>
  </div>
  <div class="Agentic AI for Wearables Data Analysis">
    <h2><a href="https://github.com/cartiktq/WearablesDataAnalysis">Agentic AI for Wearables Data Analysis</a></h2>
    <div>This codebase is a machine learning development and deployment workflow, which uses Agentic AI libraries in a microservices design, where the containerized modules
      are orchestrated by Kubernetes. This is an approximation of the work I did in one of my recent roles (private access codebase cannot be shared). 
      At a high level, the ML development worklow 
      <ul>
      <li>Collects subject physiological data such as heart rate, oxygen saturation, etc.from wearable sensors such as smart watches</li>
      <li>Performs feature engineering on the wearables data using techniques such as moving window averages and trend analysis,</li>
      <li>integrates this processed wearables data with static subject demographics and clinical history data, both sourced from different data sources,</li>
      <li>integrates with live weather data feeds which have values for temperature, air quality index, humidity, wind chill, heat index, etc.,</li>
      <li>uses a BERT encoder to generate a hyperdimensional embedding and,</li>
      <li>uses a deep learning model to predict the possible onset of adverse events such as dehydration, heat stroke, or fatigue</li> 
    </ul>
      This is a bootstrapped system that works with unlabelled data (unsupervised learning). To achieve optimal performance, the training workflow is supervised by an AI 
      agent, which sources outcomes data for the monitored subjects over time, generates the metrics such as recall and F-score from the outcomes, and uses this metric to
      set off successive training epochs until the Deep Learning NN performs at recall levels exceeding 0.9 or 0.95.

      The deployment workflow sets up an A/B deployment with a champion and a challenger ML model performing online inferences on a 70-30 split of live streaming wearables data. 

      This is an extremely simplistic scenario. It can be extended so that multiple deep learning models can be trained in parallel with a supervisory agent performing 
      hyperparameter tuning to find the best performing models. The selection of a model for deployment will be done by an agent, which oversees the A/B model deployment.   
    </div>
  </div>
  
  <div class="scRNA">
    <h2><a href="https://github.com/cartiktq/ML4SingleCellRNA">Single-Cell RNA Analysis</a></h2>
    <div>This codebase is an approximation of a consulting project I did for a biotech startup. The actual codebase is copyright protected. The Python classes
    handle specific steps including;
      <ul>
      <li>extracting single-cell RNA datasets from an open access repository such as GEO,</li>
      <li>performing preprocessing and QC on the extracted data,</li>
      <li>and the use of classical Machine Learning algorithms (implemented in scikit-learn) for biomarker identification, patient stratification, and identification of
     the specialization trajwectories of various cell types, originating from a single common progenitor, and</li>
      <li>visualization of genetic biomarkers, cell development trajectories, and patient strata using the Seaborn package</li>
    </ul>
    Each of these classes can be containerized into Docker (some of the container code is here) and invoked sequentially using an orchestrator such as Kubernetes or 
    a workflow implemented with Flyte. Actually, my work for the startup involved developing the Flyte workflow to execute this sequence of actions and deploying it on 
    an AWS Fargate cluster  
    </div>
  </div>
  
  <div class="Generative AI">
    <h2><a href="https://github.com/cartiktq/GenAI">Generative AI</a></h2>
    <p>Work with LLMs, LangChain, LangGraph, Transformers, etc. for:</p>
    <ul>
      <li>Text summarization</li>
      <li>Text generation</li>
      <li>Prompt Engineering</li>
      <li>Chain, Router, and Agent design patterns for Agentic Workflows</li>
    </ul>
  </div>
  <div class="KnowledgeGraphExploration-Java">
    <h2><a href="https://github.com/cartiktq/AcademicAnalytics">Knowledge Graph Exploration</a></h2>
    <p>The purpose of this project was to find connections between academicians - experts in mltiple disciplines, such as geology, history, law, medicine, mathematics, etc. - represented as nodes on a graph. 
      Each node is a word cloud representation of the academician with the words being sourced from publications, keynote speeches, presentations, and successful grant proposals authored and/or presented by the 
      academician. The connections between the academicians are obtained by a graph traversal algorithm, which prioritizes distinctness, as opposed to similarity between every connected pair of nodes in the path. 
      This precludes the possibility of relating academicians from the same or similar disciplines, (ex: clinicans with biologists, etc.). The publication can be found here
      (https://europepmc.org/article/med/26262370)
    </p>
  </div>
</body>
</html>
